File: DynamicProg/Triangle/Triangle.cc

class Solution {
    public:
        //From bottom to top
        int minSum(vector<vector<int>> &triangle) {
            if(triangle.empty())
                return -1;
            
            vector<vector<int>> hashmap(triangle);
            
            const int N = triangle.size();
            for(int i = 0; i < N; i++) {
                hashmap[N - 1][i] = triangle[N - 1][i];
                
            }
            
            for(int i = N - 2; i > 0; i--) {
                for(int j = 0; j < i + 1; j++ ) {
                    hashmap[i][j] = min(hashmap[i + 1][j], hashmap[i + 1][j + 1]) + triangle[i][j];
                }
            }
            return hashmap[0][0];
        }
        
        //From top to bottom
        int minSum1(vector<vector<int>> &triangle) {
            if(triangle.empty())
                return -1;
            
            vector<vector<int>> hashmap(triangle);
            
            int N = triangle.size();
            for(int i = 0; i < N; i++) {
                hashmap[N][i] = triangle[N][i];
            }
            
            for(int i = 1; i < N; ++i) {
                for(int j = 0; j < i; ++j) {
                    if(j == 0)
                        hashmap[i][j] = hashmap[i - 1][j];
                    if(j == i)
                        hashmap[i][j] = hashmap[i - 1][j - 1];
                    if((j > 0) && (j < i))
                        hashmap[i][j] = min(hashmap[i - 1][j], hashmap[i - 1][j - 1]);
                    
                    hashmap[i][j] += triangle[i][j];
                }
            }
            
            int ret = INT_MAX;
            for(int i = 0; i != N; ++i) {
                ret = min(ret, hashmap[N - 1][i]);
            }
            
            return ret;
        }
        
};

File: DynamicProg/Triangle/Triangle.java
class Solution {
    public int minSum(int[][] triangle) {
        if(triangle == null || triangle.length == 0)
            return 0;
        
        int[] last = new int[triangle.length];
        int[] cur = new int[triangle.length];
        
        last = triangle[0][0];
        cur[0] = last[0];
        
        for(int i = 1; i < triangle.length; ++i) {
            for(int j = 0; j < i + 1; ++j) {
                int sum = Integer.MAX_VALUE;
                if(j != 0)
                    sum = triangle[i][j] + last[j - 1];
                
                if(j != i)
                    sum = Math.min(sum, triangle[i][j] + last[j]);
                
                cur = sum;
            }
        }
        
        for(int k = 0; k < i + 1; k++)
            last[k] = cur[k];
        
        int min = Integer.MAX_VALUE;
        for(int n : cur)
            min = Math.min(n, min);
        
        return min;
    }
}